File: src/merchant.js
import { parseQRIS } from "./tlv.js";

export class MerchantRegistry {
  constructor() {
    this.merchants = new Map();
  }

  add(id, { qris, name }) {
    const parsed = parseQRIS(qris);
    if (!parsed.crcIsValid) throw new Error("QRIS tidak valid (CRC mismatch)");
    const m = {
      id,
      name: name || parsed.merchantName,
      qris,
      merchantName: parsed.merchantName,
      city: parsed.merchantCity,
      currency: parsed.currency,
      crcValid: parsed.crcIsValid,
    };
    this.merchants.set(id, m);
    return m;
  }

  get(id) {
    return this.merchants.get(id) || null;
  }

  list() {
    return [...this.merchants.values()].map(({ qris, ...rest }) => rest);
  }
}

File: mobile/lib/qris/merchant.dart
import 'tlv.dart';

class Merchant {
  Merchant({
    required this.id,
    this.name,
    this.merchantName,
    this.city,
    this.currency,
    this.crcValid,
  });
  final String id;
  final String? name;
  final String? merchantName;
  final String? city;
  final String? currency;
  final bool? crcValid;
}

class MerchantRegistry {
  final Map<String, Merchant> _merchants = {};

  Merchant add(String id, {required String qris, String? name}) {
    final parsed = parseQRIS(qris);
    if (!parsed.crcIsValid!) throw ArgumentError("QRIS tidak valid (CRC mismatch)");
    final m = Merchant(
      id: id,
      name: name ?? parsed.merchantName,
      merchantName: parsed.merchantName,
      city: parsed.merchantCity,
      currency: parsed.currency,
      crcValid: parsed.crcIsValid,
    );
    _merchants[id] = m;
    return m;
  }

  Merchant? get(String id) => _merchants[id];
  List<Merchant> list() => _merchants.values.toList();
}
